tests/python_package_test/test_engine.py

```diff
@@ -4509,3 +4509,27 @@ def test_quantized_training():
     quant_bst = lgb.train(bst_params, ds, num_boost_round=10)
     quant_rmse = np.sqrt(np.mean((quant_bst.predict(X) - y) ** 2))
     assert quant_rmse < rmse + 6.0
+
+
+def test_bagging_by_query_in_lambdarank():
+    rank_example_dir = Path(__file__).absolute().parents[2] / "examples" / "lambdarank"
+    X_train, y_train = load_svmlight_file(str(rank_example_dir / "rank.train"))
+    q_train = np.loadtxt(str(rank_example_dir / "rank.train.query"))
+    X_test, y_test = load_svmlight_file(str(rank_example_dir / "rank.test"))
+    q_test = np.loadtxt(str(rank_example_dir / "rank.test.query"))
+    params = {"objective": "lambdarank", "verbose": -1, "metric": "ndcg", "ndcg_eval_at": "5"}
+    lgb_train = lgb.Dataset(X_train, y_train, group=q_train, params=params)
+    lgb_test = lgb.Dataset(X_test, y_test, group=q_test, params=params)
+    gbm = lgb.train(params, lgb_train, num_boost_round=50, valid_sets=[lgb_test])
+    ndcg_score = gbm.best_score["valid_0"]["ndcg@5"]
+
+    params.update({"bagging_by_query": True, "bagging_fraction": 0.1, "bagging_freq": 1})
+    gbm_bagging_by_query = lgb.train(params, lgb_train, num_boost_round=50, valid_sets=[lgb_test])
+    ndcg_score_bagging_by_query = gbm_bagging_by_query.best_score["valid_0"]["ndcg@5"]
+
+    params.update({"bagging_by_query": False, "bagging_fraction": 0.1, "bagging_freq": 1})
+    gbm_no_bagging_by_query = lgb.train(params, lgb_train, num_boost_round=50, valid_sets=[lgb_test])
+    ndcg_score_no_bagging_by_query = gbm_no_bagging_by_query.best_score["valid_0"]["ndcg@5"]
+    print(ndcg_score_bagging_by_query, ndcg_score, ndcg_score_no_bagging_by_query)
+    assert ndcg_score_bagging_by_query >= ndcg_score - 0.1
+    assert ndcg_score_no_bagging_by_query >= ndcg_score - 0.1
```

Comment:
PR's description states that `bagging_by_query=True` should improve metrics, but I don't see any comparison of `bagging_by_query=True` and `bagging_by_query=False` here...

Reply:
Since I found the result can be random when the dataset is small. For example, on CPU, `bagging_by_query=True` gets higher NDCG with the toy test dataset (even higher than the case without bagging), while with GPU `bagging_by_query=True` could get worse results compared with `bagging_by_query=False`. But when the dataset is large, for example, with `MS LTR` dataset, the results are less random, and `bagging_by_query=True` should improve performance, as in the description of this PR.

Reply:
In addition, we also see a significant improvement in training speed with `bagging_by_query=True`.

Reply:
OK, I see. So this test is for something like "`bagging_by_query=True` doesn't break training".